src/shims/windows/thread.rs

```diff
@@ -0,0 +1,87 @@
+use rustc_middle::ty::layout::LayoutOf;
+use rustc_target::spec::abi::Abi;
+
+use crate::*;
+use shims::windows::handle::{EvalContextExt as _, Handle, PseudoHandle};
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+
+#[allow(non_snake_case)]
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn CreateThread(
+        &mut self,
+        security_op: &OpTy<'tcx, Provenance>,
+        stacksize_op: &OpTy<'tcx, Provenance>,
+        start_op: &OpTy<'tcx, Provenance>,
+        arg_op: &OpTy<'tcx, Provenance>,
+        flags_op: &OpTy<'tcx, Provenance>,
+        thread_op: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, ThreadId> {
+        let this = self.eval_context_mut();
+
+        let security = this.read_pointer(security_op)?;
+
+        // stacksize is ignored, but still needs to be a valid usize
+        let _ = this.read_scalar(stacksize_op)?.to_machine_usize(this)?;
+
+        let start_routine = this.read_pointer(start_op)?;
+
+        let func_arg = this.read_immediate(arg_op)?;
+
+        let flags = this.read_scalar(flags_op)?.to_u32()?;
+
+        let thread = if this.ptr_is_null(this.read_pointer(thread_op)?)? {
+            None
+        } else {
+            let thread_info_place = this.deref_operand(thread_op)?;
+            Some(thread_info_place)
+        };
+
+        let stack_size_param_is_a_reservation =
+            this.eval_windows("c", "STACK_SIZE_PARAM_IS_A_RESERVATION")?.to_u32()?;
+
+        if flags != 0 && flags != stack_size_param_is_a_reservation {
+            throw_unsup_format!("unsupported `dwCreationFlags` {} in `CreateThread`", flags)
+        }
+
+        if !this.ptr_is_null(security)? {
+            throw_unsup_format!("non-null `lpThreadAttributes` in `CreateThread`")
+        }
+
+        this.start_thread(
+            thread,
+            start_routine,
+            Abi::System { unwind: false },
+            func_arg,
+            this.layout_of(this.tcx.types.u32)?,
+        )
+    }
+
+    fn WaitForSingleObject(
+        &mut self,
+        handle_op: &OpTy<'tcx, Provenance>,
+        timeout_op: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let handle = this.read_scalar(handle_op)?.check_init()?;
+
+        let timeout = this.read_scalar(timeout_op)?.to_u32()?;
+
+        let thread = match Handle::from_scalar(handle, this)? {
+            Some(Handle::Thread(thread)) => thread,
+            // Unlike on posix, joining the current thread is not UB on windows.
```

Comment:
I can't find any docs explicitly stating what happens when a thread waits for itself to end. But because it isn't listed as UB anywhere, it logically can only be a deadlock. When testing on real windows, it does indeed deadlock.

Reply:
One could also say if it's not explicitly allowed anywhere it could be UB, and that might just happen to be a deadlock right now...

... but a library function having UB in such a situation without calling it out does seem unlikely. So, seems like a fair assumption to me, but please state that in the code via comments.